text/0023-starlark_load_dataset.md

```diff
@@ -0,0 +1,249 @@
+
+# require that the permissions of the dataset that results from this transform
+# match the permissions of "b5/city_residents"
+residents = load_dataset("b5/city_residents", forward_permissions=True)
+```
+
+These are suggestions only. We need lots of time to think through how these options will work, this rfc only adds the assumption that those statements will be expressed as options on `load_dataset`. We can build on the already-present assumption that loaded datasets are read-only, making all options pertain to read access.
+
+We _do_ need to declare what the default `load_dataset` expression means in relation to permissions:
+
+```python
+residents = load_dataset("b5/city_residents")
+```
+
+The default load dataset should be taken to mean "load this dataset with the _maximum available permission_". If a dataset is open, the user is free to read from any and all parts of the document.
+
+Combining a default of requesting maximum permission with _restrictive_ permission declared on the imported dataset supports our default stance of an open dataset ecosystem. The onus of declaring how a dataset can be used will fall to the owner of the dataset, and defaulting to max permissions forces the owner to assume that "if it's permitted, someone will use it", which is the correct assumption when attempting to restrict access.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`load_dataset` has a fairly simple definition:
+
+```
+def load_dataset(reference): # return: dataset object
+```
+
+* **reference** is a string reference to a dataset, as defined by our dataset naming conventions elsewhere.
+* **dataset object** the returned value is a dataset object, as defined by ds.NewDataset in `qri-io/startf/ds`.
+
+As mentioned earler, the plan is to expand this definition with options once we've had time to do more research. Theses examples of valid reference strings, with `QmProfile` and `QmHash` being standins for full base-58 encoded hashes:
+
+```python
+load_dataset("b5/dataset")
+load_dataset("b5/dataset@QmProfile/ipfs/QmHash")
+load_dataset("@/ipfs/QmHash")
+```
+
+### `load_dataset` must be top level scope
+`load_dataset` statements will only be allowed within the primary scope of a transform script. Phrased another way: `load_dataset` _cannot_ be called within a function. This limits the scope of what is possible when composing dynamic import statements. This **won't** work:
+
+```python
+load("qri.star", "qri")
+
+# create a dataset of all dataset titles
+def transform(ds, ctx):
+  # error: load_dataset must be called in top level scope
+  datasets = [load_dataset(name) for name in qri.list_datasets()]
+  ds.set_body([ds.get_meta('title') for ds in datasets])
+```
+
+Statements that use scropts to compose import names, however, can and should work:
+
+```python
+pop_years = [load_dataset("b5/population_%s" % year) for year in range(2011,2017)]
```

Comment:
I believe that we shouldn't allow this. It breaks one of our earlier stated goals: explicit presentation of dependencies. In fact, I think that all arguments passed to "load_dataset" must be literal strings. I don't know if starlark allows us to check this, perhaps we need to parse the source ourselves?

Reply:
agreed. Not sure how we're going to pull this off, but let's make it a design goal & see how hard starlark makes our lives.